src/webui/webapplication.cpp

```diff
@@ -88,6 +90,66 @@ namespace
         } \
     }
 
+    const QMap<QString, QString> APICompatMapping1 {
```

Comment:
They has "1"  and "2" in the names, suggesting that others might follow. Then what about possible parameter changes? Also, the old object paths denote possibility to export several objects. Quite a valuable possibility!

Reply:
> the old object paths denote possibility to export several objects

For example, you could export the RSS subsytem in its own sub-path. While now everything is at the same level potentially creating a mess.

P.S. Take a look at QDBusAbstractAdaptor documentation and http://doc.qt.io/qt-5/usingadaptors.html to get an idea what Web API implementation might look like.

Reply:
>They has "1" and "2" in the names, suggesting that others might follow. Then what about possible parameter changes? Also, the old object paths denote possibility to export several objects. Quite a valuable possibility!

@evsh, I can't say I understand what you mean. This is only "old-name to new-name" mapping tables. I have two since they are used in a different way (because of old API inconsistency).
Or are you just looking to the future, and you want to use something like that for the different things? In this case, do not be lazy to give detailed examples of expected use cases.
Will reading QDBusAbstractAdaptor documentation be enough to understand you?

Reply:
Yes, this is about future developments (although API sub-paths would make sense in any case). The idea behind QDBusAbstractAdaptor is very simple: you provide a QObject and a path to publish it, and then all its slots and properties (here you additionally mark them as PUBLIC_API and PRIVATE_API, but that doesn't matter) become available at that path. Qt DBus documentation can explain details.

It would be nice to achieve the same for HTTP, especially taking into account that you are already only a step away.

Reply:
Sorry, forgot to comment on the first part.

> @evsh, I can't say I understand what you mean. This is only "old-name to new-name" mapping tables. I have two since they are used in a different way (because of old API inconsistency).

I'm thinking of possible parameters change between APIs. The maps we are talking about might contain not name → name pairs, but name → function to handle parameter change. Or such changes in the API are impossible or quite unlikely?

Reply:
Quick follow-up: supposing the proposed object publishing way is adopted, we could just publish an object for the old API and avoid new on old sub-system dependency (look and the hash parameter handling in the current PR code).

Reply:
>you provide a QObject and a path to publish it, and then all its slots and properties (here you additionally mark them as PUBLIC_API and PRIVATE_API, but that doesn't matter) become available at that path. 

Hmm... Interesting idea. I'll try to go there...
>although API sub-paths would make sense in any case

Agree. As I said before I got rid from sub-paths since they're currently confused/messed and I went a short ways (no "subpaths" is better than wrong "subpaths"). However, now I wonder to use it again.
However, I don't want to delay thi PR, because there are other goals. Due to the fact that the compatibility layer is implemented, we can have a old API as main and new one as experimental feature for some time. Then we can change the API paths safe for the end user. What do you say?
I'm just not sure I can quite quickly develop the consistent API paths and names. However, you @Chocobo1, you too) can help me with this. You can post your suggestions in this PR topic.

Reply:
Everything is clear with the RSS, of course: `/api/rss/processingEnabled` and so on.

Reply:
>I'm just not sure I can quite quickly develop the consistent API paths and names. However, you @Chocobo1, you too) can help me with this. You can post your suggestions in this PR topic.

You are talking about sub-paths under `/api` right?
Take a look, WebAPIs is not much: https://github.com/qbittorrent/qBittorrent/wiki/WebUI-API-Documentation
I can only think of `/api/torrent`: for APIs fetching single torrent data, `/api/global`: for APIs like "Get global download limit" and everything else still under `/api` because no other sub-path fits.

>Then we can change the API paths safe for the end user.

I don't think end users cares much about it (as long as webUI still work), it's 3rd-party software developers whom is related to these changes: https://github.com/SickGear/SickGear/issues/979#issuecomment-341169007
I'm not saying that is the golden rule, but we something we should bear in mind.

Reply:
Maybe `api/session` for bits like the global download limit? Login/logout might also fit into the session sub-object.

Reply:
>Login/logout might also fit into the session sub-object.

sound redundant, users are logging into a web session, not a libtorrent session.

Reply:
Agree. From the WebAPI POV Session is Web session. Globals can be under `app/` (or some other) subpath.

Reply:
Obvious remark: if we are going to move to object publishing approach (see above) we need a sub-path for each API entry.

Reply:
>if we are going to move to object publishing approach

Not so fast, I think... I tried to imagine it myself, and I have a lot of questions now (and very little time for this).
So I plan to implement API paths another (more simple) way in this PR so as not to delay it. Are there any strong objections against this?

Reply:
Hi, a note from users of this API [here](https://github.com/qbittorrent/qBittorrent/pull/7610#issuecomment-342182931)

Reply:
> Are there any strong objections against this?

No objections at all.

Reply:
@JackDandy:  currently this PR does provide complete compatibility with the old AP (except maybe for a version string, I'm not sure about that). Thanks for your concerns!